lib/schemas/rfc7239.js

```diff
@@ -0,0 +1,54 @@
+'use strict'
+
+function splitMap (string, separator, cb) {
+  // split into elements
+  return string.split(separator)
+    .filter(Boolean)
+    .forEach(cb)
+}
+
+function parsePart (part) {
+  var pair = part.split(/ *= */)
+
+  var name = pair[0].toLowerCase()
+  var value = pair[1]
+
+  if (value) {
+    switch (typeof this[name]) {
+      case 'undefined':
+        this[name] = value
+        break
+
+      // convert to array
+      case 'string':
+        this[name] = [this[name], value]
+        break
+
+      // append to array
+      case 'object':
+        this[name].push(value)
+        break
+    }
+  }
+}
+
+var ELEMENT_SEPARATOR = / *; */
+var PART_SEPARATOR = / *, */
+
+module.exports = function (req) {
+  var forwarded = {}
+  var header = req.headers.forwarded
+
+  if (!header) {
+    return forwarded
+  }
+
+  splitMap(header, ELEMENT_SEPARATOR, function parseElement (el) {
```

Comment:
Basically, what I'm saying is that the current parse is actually _not RFC 7239 compliant_, as you are not correctly following the ABNF, which allows for `,` and `;` to appear within a `quoted-string` and the defined ABNF for the header is as follows:

```
Forwarded   = 1#forwarded-element
forwarded-element = [ forwarded-pair ] *( ";" [ forwarded-pair ] )
forwarded-pair = token "=" value
value          = token / quoted-string
token          = 1*tchar
tchar          = "!" / "#" / "$" / "%" / "&" / "'" / "*"
               / "+" / "-" / "." / "^" / "_" / "`" / "|" / "~"
               / DIGIT / ALPHA
quoted-string  = DQUOTE *( qdtext / quoted-pair ) DQUOTE
qdtext         = HTAB / SP /%x21 / %x23-5B / %x5D-7E / obs-text
obs-text       = %x80-FF
```

The given header, `Forwarded: for="foo,bar"` is syntactically valid, yet this module will definitely parse it incorrectly.

Because the RFC allows extensions, perhaps your proxy decides it's going to include a custom field of the authenticated user's name. Uh, oh!
```
# this is in bash, so the "" is an escaped double quote (i.e. it's only a single double quote in the program)
$ node -pe "require('forwarded')({connection:{remotePort:0},headers:{forwarded:'user=""bad,for=_whitelisted_ip"";for=_blacklisted_ip'}})"
{ addrs: [ 'whitelisted_ip', 'blacklisted_ip' ],
  by: null,
  host: undefined,
  port: '0',
  ports: [ '0' ],
  proto: 'http' }
```

Even though the header is 100% syntactically valid, with a single forwarded entry of `{for: '_blacklisted_ip', user: 'bad,for=_whitelisted_ip'}`, we end up thinking that the proxy got the request from `_whitelisted_ip` incorrectly due to the invalid parsing this module did :(

And I'm not even a security expert; I can tell you that if we don't fix the parsing, it will most likely end up as a critical security bug.